File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
       
        if(nums1.size()>nums2.size())return findMedianSortedArrays(nums2,nums1);
        
        int l1=nums1.size(),l2=nums2.size();
        int tl=l1+l2;
        int low=0,high=l1-1;
        double leftx,rightx,lefty,righty;
        while(low-high<=1)
        {
            int partx=low+(high-low)/2;
            int party=(tl+1)/2-partx;
            
            double leftx= (partx==0)?INT_MIN:nums1[partx-1];
            double rightx= (partx==l1)?INT_MAX:nums1[partx];
            double lefty= (party==0)?INT_MIN:nums2[party-1];
            double righty=(party==l2)?INT_MAX:nums2[party];
            //cout<<leftx<<" "<<rightx<<" "<<lefty<<" "<<righty<<endl;
            if(leftx<=righty&&lefty<=rightx)
            { 
                // cout<<leftx<<" "<<rightx<<" "<<lefty<<" "<<righty<<endl;
                if(tl%2)return max(leftx,lefty);
                else return (max(leftx,lefty)+min(rightx,righty))/2;
            }
            else if(leftx>righty)
            {
                high=partx-1;
            }
            else
            {
                low=partx+1;
            }
        }
        
       
        return 0.0;
       
                     
    }
};

File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int l1 = nums1.length, l2 = nums2.length;
        int totalLen = l1 + l2;

        int l = -1, r = l1;

        while (r - l > 1) {
            int m = l + (r - l) / 2;
            int partX = m;
            int partY = (totalLen + 1) / 2 - (partX+1) - 1; 
            // Explanation:
            // nums1's left partition is [0..partX], so left max at nums1[partX]
            // nums2's left partition is [0..partY], note offset by -2 due to zero indexing adjustment

            double leftX = (partX < 0) ? Integer.MIN_VALUE : nums1[partX];
            double rightX = (partX + 1 >= l1) ? Integer.MAX_VALUE : nums1[partX + 1];

            double leftY = (partY < 0) ? Integer.MIN_VALUE : nums2[partY];
            double rightY = (partY + 1 >= l2) ? Integer.MAX_VALUE : nums2[partY + 1];

            if (leftX <rightY) {
                l = m;
            } else {
                r = m;
            }
        }

        int partX = l;
        int partY = (totalLen + 1) / 2 - partX - 2;

        double leftX = (partX < 0) ? Integer.MIN_VALUE : nums1[partX];
        double rightX = (partX + 1 >= l1) ? Integer.MAX_VALUE : nums1[partX + 1];

        double leftY = (partY < 0) ? Integer.MIN_VALUE : nums2[partY];
        double rightY = (partY + 1 >= l2) ? Integer.MAX_VALUE : nums2[partY + 1];

        if ((totalLen & 1) == 1) {
            return Math.max(leftX, leftY);
        } else {
            return (Math.max(leftX, leftY) + Math.min(rightX, rightY)) / 2.0;
        }
    }
}
